pennylane/ops/projection.py

```diff
@@ -0,0 +1,85 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the available built-in quantum projection operations
+supported by PennyLane.
+"""
+
+import numpy as np
+
+from pennylane.operation import AnyWires, Projection
+
+
+class Measure(Projection):
+    r"""Measure(wires)
+    A measurement in the computational basis.
```

Comment:
I have to admit, reading further down, I'm getting confused myself! @trbromley, maybe you also have thoughts here.

So far, we have four measurement concepts/objects in PL:

1. `Projection`, an abstract base class that inherits from `Operation` (this makes sense to me).

2. `Measure`, which inherits from `Projection`.

3. Terminal measurement statistics, `qml.expval(..)`, `qml.var(..)`, etc. These have several roles:

    - They process the arguments, and return a `MeasurementProcess` instance, that contains all the information about the measurement.
    - They provide a user-interface to the user, without the user needing to create a measurement process object themselves. However, it _is_ possible to use a `MeasurementProcess` directly if you want, e.g., `return qml.MeasurementProcess(obs=qml.PauliZ(0))`

4. The `MeasurementProcess` class, which:

   - Contains information about the measurement (observables, eigenvalues, wires, etc).
   - Interacts with PennyLane's queuing system, to make sure that the process is added to the QNode.

To me, (1) makes sense.

However (2) confuses me - it's a specific implementation of the abstract `Projection` class, but it has a generic `Measure` name, which could also be confused with the existing (but separate) terminal measurement system!

Reply:
Something like the following, instead, makes a lot more sense from my point of view:

```python
@qml.qnode(dev)
def circuit():
    qml.Hadamard(wires=0)
    qml.CNOT(wires=[0, 1])
    qml.ProjectiveMeasure(projector=?, basis=qml.PauliX, wires=[0, 1])
    return [qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))]
```

* Having a more specific name (`ProjectiveMeasure`) reduces potential confusion, and emphasizes mid-circuit measurement.

* I would make `ProjectiveMeasure` a function, that returns a `MeasurementProcess` object with information about the projective measurement the device needs to undertake. That is, it should go in the `pennylane/tape/measure.py`, and modifications made to `MeasurementProcess` so that it can represent a projective measurement. This also avoids the need for having to subclass `Projector` every time you want to define a new projector; here, it is simply passed as a keyword argument.

I'm curious to get everyone's thoughts (@trbromley, @ixfoduap, @Lucaman99).

Reply:
Hi @josh146,

Thanks for the suggestions and thoughtfulness. I have to say, I'm not a big fan. I understand that there is some potential for confusion, but in my opinion this is minor. There are terminal measurement statistics, which are handled by ``MeasurementProcess``, and mid-circuit measurements that are handled by  ``Measure()``. From a user perspective, the main change that appears is that PennyLane now supports mid-circuit measurements, which are in a category of their own. Great!

I also prefer a UI with ``Measure()``, which is cleaner
```python
@qml.qnode(dev)
def circuit():
    qml.Hadamard(wires=0)
    qml.CNOT(wires=[0, 1])
    qml.Measure(wires=[0, 1])
    return [qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))]
```

Finally, we shouldn't forget everything that was discussed in the [ADR for conditional measurements](https://github.com/PennyLaneAI/adrs/blob/master/documents/041-conditional_operations.rst). I'm concerned that if we start deviating from that plan, we'll find ourselves having to deal with how these changes spread across the proposed architecture.

In short, I prefer to keep as is and just be extra careful in the documentation and tutorials to avoid potential confusions

Reply:
Yes, my main worry is if a user tries to do: `return qml.Measure(wires=range(2))`. Indeed, this seems like a very natural thing to do and I would do it and be surprised that it didn't work. `Measure` seems to suggest extracting some classical information from the system. I think there will be less confusion with concepts like `Projection`, as it sounds less like extracting information from the system and more like performing an operation on it.

For the sake of following the ADR, could we just rename `class Measure(Projection):` to `class Project(Projection):`
Since it is projecting into the qubit computational basis, it can have a very general name. Other projections could be named, e.g., `ProjectPlusMinus`.

Reply:
> For the sake of following the ADR

I just want to emphasize that we do not need to adhere to an ADR simply because it was accepted! ADRs are not set in stone, and we can always revisit if we have new ideas or information. In fact, there have been quite a few cases in the past where reviewer comments during the implementation have led to (good) design changes.

If it becomes clear during the implementation of an already accepted ADR that there are

* unexpected problems, or
* new issues, or
* even if it becomes clear that there is a better way of designing/implementing,

it is perfectly fine to discuss and revise.

Regarding this comment thread specifically, my concern is less with the name and more in terms of the implementation; in particular, the need to create operation subclasses simply for defining new projectors.

With the new core being default, we have a plan to update the devices to understand the `MeasurementProcess` natively. The `MeasurementProcess` class already supports measurements defined via eigenvalues; would it not be still in scope to allow it, or a subclass of it, to represent projectors as well?

The

Reply:
It seems like we need to have chat! Some friendly counterarguments:

- I mentioned the context of the ADR not because it means things are set in stone, but rather to make sure we keep all the pieces in mind. Ultimately we want to be able to use mid-circuit measurements to do several tasks. For example we want to do something like ``b = qml.Measure(wires=0)``, then post-porcess `b` and pass it to a subsequent operation. Also we want to interface with a new `qml.if()`. It took us month to see how all the pieces fit together and I worry that making one local change here takes us away from the global minimum
- I personally really like to think of `Measure()` as an operation: I can rotate the qubit, entangle it with a qubit, or measure it. Perhaps I need to understand `MeasurementProcess` better, but to me it seems like an overkill. Also, would it allows us to specify instructions in the simulators on how to update the state after a measurement has occured?
- The projection class was added as a suggestion from Nathan due to the potential usefulness of performing different kinds of projections in different contexts. We can definitely revisit this with new eyes
- I guess I would just like to see a clearer example of how we can use `MeasurementProcess` to support the same intended functionality that is done with this PR and that fits the bigger picture of supporting conditional operations (dynamical circuits) in PennyLane. This is unfortunately not very clear to me